wgpu-core/src/command/mod.rs

```diff
@@ -78,21 +78,185 @@ pub(crate) enum CommandEncoderStatus {
+            Self::Error => Err(CommandEncoderError::Invalid),
+        }
+    }
+
+    /// Locks the encoder by putting it in the [`Self::Locked`] state.
+    ///
+    /// Call [`Self::unlock_encoder`] to put the [`CommandBuffer`] back into the [`Self::Recording`] state.
+    fn lock_encoder(&mut self) -> Result<(), CommandEncoderError> {
+        match *self {
+            Self::Recording(_) => {
+                change_state_with_inner!(self, Self::Recording, Self::Locked);
+                Ok(())
+            }
+            Self::Finished(_) => Err(CommandEncoderError::NotRecording),
+            Self::Locked(_) => {
+                let _ = mem::replace(self, Self::Error);
+                Err(CommandEncoderError::Locked)
+            }
+            Self::Error => Err(CommandEncoderError::Invalid),
+        }
+    }
+
+    /// Unlocks the [`CommandBuffer`] and puts it back into the [`Self::Recording`] state.
+    ///
+    /// This function is the unlocking counterpart to [`Self::lock_encoder`].
+    ///
+    /// It is only valid to call this function if the encoder is in the [`Self::Locked`] state.
+    fn unlock_encoder(&mut self) -> Result<EncoderGuard<'_>, CommandEncoderError> {
+        match *self {
+            Self::Locked(_) => {
+                change_state_with_inner!(self, Self::Locked, Self::Recording);
+                Ok(EncoderGuard {
+                    inner: self,
+                    succeeded: false,
+                })
+            }
+            Self::Finished(_) => Err(CommandEncoderError::NotRecording),
+            Self::Recording(_) => {
+                let _ = mem::replace(self, Self::Error);
+                Err(CommandEncoderError::Invalid)
+            }
+            Self::Error => Err(CommandEncoderError::Invalid),
+        }
+    }
+
+    fn finish(&mut self, device: &Device) -> Result<(), CommandEncoderError> {
+        match self {
+            Self::Recording(inner) => {
+                if let Err(e) = inner.encoder.close(device) {
+                    let _ = mem::replace(self, Self::Error);
+                    Err(e.into())
+                } else {
+                    change_state_with_inner!(self, Self::Recording, Self::Finished);
+                    // Note: if we want to stop tracking the swapchain texture view,
+                    // this is the place to do it.
+                    Ok(())
+                }
+            }
+            Self::Finished(_) => Err(CommandEncoderError::NotRecording),
+            Self::Locked(_) => {
+                let _ = mem::replace(self, Self::Error);
+                Err(CommandEncoderError::Locked)
+            }
+            Self::Error => Err(CommandEncoderError::Invalid),
+        }
+    }
+}
+
+/// A guard on which [`EncoderGuard::succeeded`] should be called if
+/// no errors have been raised during encoding, otherwise the encoder will be put in the
+/// [`CommandEncoderStatus::Error`] on [`EncoderGuard::drop`].
```

Comment:
I think it would be nice to highlight the dereferencing behavior, too. Assuming we change `succeeded` to take `self` by value:

```suggestion
/// A guard to enforce error reporting, for a [`CommandBuffer`] in the [`Recording`] state.
///
/// An `EncoderGuard` holds a mutable reference to a [`CommandEncoderStatus`] that
/// has been verified to be in the [`Recording`] state. The `EncoderGuard` dereferences
/// mutably to the [`CommandBufferMutable`] that the status holds.
///
/// Dropping an `ErrorGuard` sets the [`CommandBuffer`]'s state to
/// [`CommandEncoderStatus::Error`]. If your use of the guard was
/// successful, call its [`succeeded`] method to dispose of it.
///
/// [`Recording`]: CommandEncoderStatus::Recording
```